### DISCUSSION TITLE
can strogswan use all cpu cores in single SA ipsec connection

### DISCUSSION THREAD

**Discussion by alireza11048** [strongswan/strongswan#1089] [Q&A] [answered]
Hi,
I've used strongswan as ipsec engine between two device, which only one ipsec connection (single SA) is stablished between the devices. in the htop command output, i can see that only one cpu's core is used to encrypt/decrypt the packets. is this possible to distribute the load between cores to increase the bandwith?
I've seen https://wiki.strongswan.org/projects/strongswan/wiki/Pcrypt, but when i enter the "modprobe tcrypt alg="pcrypt(authenc(hmac(sha256),ctr(aes)))" type=3 mode=500"  command, i got the below error:

> modprobe: ERROR: could not insert 'tcrypt': Unknown symbol in module, or unknown parameter (see dmesg)
and in the dmesg the below error apears:

> [99921.070403] tcrypt: one or more tests failed!

thanks in advance

**Comment by tobiasbrunner** [comment, marked as answer]
As documented, that command is expected to fail (not sure if that error is an actual problem or not). The module is quite old and I don't think its use is recommended anymore. Anyway, the Linux kernel doesn't really support using multiple cores for single SAs/flows.

Also, if you want more performance, use AES-GCM and not AES-CTR/CBC and an HMAC.

**Reply by alireza11048** [reply]
thanks for your reply,
by indicating leftprotoport and rightprotoport in the ipsec conf, to create different SA for different ports, can i expect higher throughput as the crypto process for different SA/flow will be done on different cpu cores?

if yes, is there another way to create multiple SAs, except than leftprotoport configuration?

thanks in advance

**Reply by tobiasbrunner** [reply]
It might but there is no guarantee. In particular, for inbound traffic it depends on e.g. whether your network interface has multiple queues and whether you can assign SAs to them and how these are then mapped to CPU cores (RSS etc.).

For outbound traffic there might be other ways than proto/port specific policies to direct traffic to specific SAs (and maybe CPU cores, not sure how the kernel handles that exactly, though), e.g. marks or XFRM interfaces.

For a while we also worked on a [draft to create per-CPU SAs](https://datatracker.ietf.org/doc/html/draft-pwouters-ipsecme-multi-sa-performance) (experimental implementations are mentioned in the draft), but that didn't get much traction.

**Reply by jlearman** [reply]
My testing on Ubuntu 22.04 on AWS `c5n.2xlarge` with `aesni_intel` loaded and without hardware IPsec acceleration shows that for traffic being encrypted, CPU load is distributed evenly across CPUs, regardless of the number of `ip xfrm state` entries (child SAs) or how ECMP is configured.

However, for traffic being decrypted, all traffic is assigned to the same CPU, in all ECMP configurations.

The ECMP configurations I've tried:
- /proc/sys/net/ipv4/fib_multipath_hash_policy=0 (L3 ECMP)
- /proc/sys/net/ipv4/fib_multipath_hash_policy=1 (L4 ECMP)
- /proc/sys/net/ipv4/fib_multipath_hash_policy=4 (Custom ECMP), /proc/sys/net/ipv4/fib_multipath_hash_fields=255 (all fields)

I believe that this is Linux kernel behavior which is not controlled by Strongswan per se.  But I was hoping that someone had suggestions on how to spread the load for traffic being decrypted.

Note that each SA has state information to avoid replay attacks, so decryption for any given SA should be assigned to a single CPU, to avoid the need for atomics when accessing this state information.
